File: src/main/java/in/dragonbra/javasteam/types/KVTextReader.java
package in.dragonbra.javasteam.types;

import in.dragonbra.javasteam.util.Passable;
import in.dragonbra.javasteam.util.Strings;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.PushbackInputStream;
import java.nio.charset.StandardCharsets;
import java.util.Collections;
import java.util.Map;
import java.util.TreeMap;

/**
 * @author lngtr
 * @since 2018-02-26
 */
public class KVTextReader extends PushbackInputStream {

    public static final Map<Character, Character> ESCAPED_MAPPING;

    static {
        Map<Character, Character> escapedMapping = new TreeMap<>();

        escapedMapping.put('n', '\n');
        escapedMapping.put('r', '\r');
        escapedMapping.put('t', '\t');
        escapedMapping.put('\\', '\\');

        ESCAPED_MAPPING = Collections.unmodifiableMap(escapedMapping);
    }

    KVTextReader(KeyValue kv, InputStream is) throws IOException {
        super(is);
        Passable<Boolean> wasQuoted = new Passable<>(false);
        Passable<Boolean> wasConditional = new Passable<>(false);

        KeyValue currentKey = kv;

        do {
            String s = readToken(wasQuoted, wasConditional);

            if (Strings.isNullOrEmpty(s)) {
                break;
            }

            if (currentKey == null) {
                currentKey = new KeyValue(s);
            } else {
                currentKey.setName(s);
            }

            s = readToken(wasQuoted, wasConditional);

            if (wasConditional.getValue()) {
                // Now get the '{'
                s = readToken(wasQuoted, wasConditional);
            }

            if (s.startsWith("{") && !wasQuoted.getValue()) {
                // header is valid so load the file
                currentKey.recursiveLoadFromBuffer(this);
            } else {
                throw new IllegalStateException("LoadFromBuffer: missing {");
            }

            currentKey = null;
        } while (!endOfStream());
    }

    private void eatWhiteSpace() throws IOException {
        while (!endOfStream()) {
            if (!Character.isWhitespace((char) peek())) {
                break;
            }

            read();
        }
    }

    private boolean eatCPPComment() throws IOException {
        if (!endOfStream()) {
            char next = (char) peek();

            if (next == '/') {
                readLine();
                return true;
                /*
                 *  As came up in parsing the Dota 2 units.txt file, the reference (Valve) implementation
                 *  of the KV format considers a single forward slash to be sufficient to comment out the
                 *  entirety of a line. While they still _tend_ to use two, it's not required, and likely
                 *  is just done out of habit.
                 */
            }

            return false;
        }
        return false;
    }

    private void readLine() throws IOException {
        char c;
        do {
            c = (char) read();
        } while (c != '\n' && !endOfStream());
    }

    private byte peek() throws IOException {
        int p = read();
        if (p >= 0) {
            unread(p);
        }
        return (byte) p;
    }

    public String readToken(Passable<Boolean> wasQuoted, Passable<Boolean> wasConditional) throws IOException {
        wasQuoted.setValue(false);
        wasConditional.setValue(false);

        while (true) {
            eatWhiteSpace();

            if (endOfStream()) {
                return null;
            }

            if (!eatCPPComment()) {
                break;
            }
        }

        if (endOfStream()) {
            return null;
        }

        char next = (char) peek();
        if (next == '"') {
            wasQuoted.setValue(true);

            // "
            read();

            ByteArrayOutputStream baos = new ByteArrayOutputStream();

            while (!endOfStream()) {
                if (peek() == '\\') {
                    read();

                    char escapedChar = (char) read();

                    Character replacedChar = ESCAPED_MAPPING.get(escapedChar);
                    if (replacedChar == null) {
                        replacedChar = escapedChar;
                    }

                    baos.write(replacedChar);

                    continue;
                }

                if (peek() == '"') {
                    break;
                }

                baos.write(read());
            }

            // "
            read();

            // convert the output stream as an utf-8 supported string.
            return baos.toString(StandardCharsets.UTF_8);
        }

        if (next == '{' || next == '}') {
            read();
            return String.valueOf(next);
        }

        boolean bConditionalStart = false;
        int count = 0;
        StringBuilder ret = new StringBuilder();

        while (!endOfStream()) {
            next = (char) peek();

            if (next == '"' || next == '{' || next == '}') {
                break;
            }

            if (next == '[') {
                bConditionalStart = true;
            }

            if (next == ']' && bConditionalStart) {
                wasConditional.setValue(true);
            }

            if (Character.isWhitespace(next)) {
                break;
            }

            if (count < 1023) {
                ret.append(next);
            } else {
                throw new IOException("ReadToken overflow");
            }

            read();
        }
        return ret.toString();
    }

    private boolean endOfStream() {
        try {
            return peek() == -1;
        } catch (IOException e) {
            return true;
        }
    }
}

File: src/main/java/in/dragonbra/javasteam/types/KVTextReader.kt
package `in`.dragonbra.javasteam.types

import `in`.dragonbra.javasteam.util.Passable
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.lang.IllegalStateException
import java.lang.StringBuilder
import java.nio.charset.StandardCharsets

/**
 * @author lngtr
 * @since 2018-02-26
 */
class KVTextReader
@Throws(IllegalStateException::class, IOException::class)
internal constructor(
    kv: KeyValue,
    inputStream: InputStream,
) : InputStreamReader(inputStream, StandardCharsets.UTF_8) {

    companion object {
        @JvmField
        val ESCAPED_MAPPING = mapOf<Char, Char>(
            '\\' to '\\',
            'n' to '\n',
            'r' to '\r',
            't' to '\t',
            // todo: (SK) any others?
        )
    }

    private val sb = StringBuilder(128)

    private var peekedChar: Int? = null

    // Mimics C# StreamReader 'Peek()'
    val peek: Int
        get() {
            if (peekedChar == null) {
                peekedChar = read()
            }
            return peekedChar ?: -1
        }

    // Mimics C# StreamReader 'EndOfStream'
    val endOfStream: Boolean
        get() {
            return try {
                peek == -1
            } catch (_: IOException) {
                true
            }
        }

    init {
        val wasQuoted = Passable(false)
        val wasConditional = Passable(false)

        var currentKey: KeyValue? = kv

        do {
            var s = readToken(wasQuoted, wasConditional)

            if (s.isNullOrEmpty()) {
                break
            }

            if (currentKey == null) {
                currentKey = KeyValue(s)
            } else {
                currentKey.name = s
            }

            s = readToken(wasQuoted, wasConditional)

            if (wasConditional.value == true) {
                // Now get the '{'
                s = readToken(wasQuoted, wasConditional)
            }

            if (s != null && s.startsWith("{") && wasQuoted.value == false) {
                // header is valid so load the file
                currentKey.recursiveLoadFromBuffer(this)
            } else {
                throw IllegalStateException("LoadFromBuffer: missing {")
            }

            currentKey = null
        } while (!endOfStream)
    }

    // override read() to peek the char.
    override fun read(): Int {
        if (peekedChar != null) {
            val result = peekedChar!!
            peekedChar = null
            return result
        }
        return super.read()
    }

    @Throws(IOException::class)
    private fun eatWhiteSpace() {
        while (!endOfStream) {
            if (!peek.toChar().isWhitespace()) {
                break
            }

            read()
        }
    }

    @Throws(IOException::class)
    private fun eatCPPComment(): Boolean {
        if (!endOfStream) {
            val next = peek.toChar()

            if (next == '/') {
                readLine()
                return true
                /*
                 *  As came up in parsing the Dota 2 units.txt file, the reference (Valve) implementation
                 *  of the KV format considers a single forward slash to be sufficient to comment out the
                 *  entirety of a line. While they still _tend_ to use two, it's not required, and likely
                 *  is just done out of habit.
                 */
            }

            return false
        }

        return false
    }

    @Throws(IOException::class)
    private fun readLine() {
        var c: Char
        do {
            c = read().toChar()
        } while (c != '\n' && !endOfStream)
    }

    @Throws(IOException::class)
    fun readToken(wasQuoted: Passable<Boolean>, wasConditional: Passable<Boolean>): String? {
        wasQuoted.value = false
        wasConditional.value = false

        while (true) {
            eatWhiteSpace()

            if (endOfStream) {
                return null
            }

            if (!eatCPPComment()) {
                break
            }
        }

        if (endOfStream) {
            return null
        }

        var next = peek.toChar()
        if (next == '"') {
            wasQuoted.value = true

            // "
            read()

            sb.clear()
            while (!endOfStream) {
                if (peek.toChar() == '\\') {
                    read()

                    val escapedChar = read().toChar()
                    val replacedChar = ESCAPED_MAPPING[escapedChar] ?: escapedChar

                    sb.append(replacedChar)

                    continue
                }

                if (peek.toChar() == '"') {
                    break
                }

                sb.append(read().toChar())
            }

            // "
            read()

            return sb.toString()
        }

        if (next == '{' || next == '}') {
            read()
            return next.toString()
        }

        var bConditionalStart = false
        val count = 0
        sb.clear()
        while (!endOfStream) {
            next = peek.toChar()

            if (next == '"' || next == '{' || next == '}') {
                break
            }

            if (next == '[') {
                bConditionalStart = true
            }

            if (next == ']' && bConditionalStart) {
                wasConditional.value = true
            }

            if (next.isWhitespace()) {
                break
            }

            // count isn't used anymore, but still defined in SK.
            @Suppress("KotlinConstantConditions")
            if (count < 1023) {
                sb.append(next)
            } else {
                throw IOException("ReadToken overflow")
            }

            read()
        }

        return sb.toString()
    }
}
